go/vumitools/contact/tests/test_models.py

```diff
@@ -278,3 +279,341 @@ def test_new_contact_for_addr_gtalk(self):
+                groups=[group])
+            contact_keys.add(contact.key)
+
+        index_page = yield store.get_static_contact_keys_for_group(group)
+        self.assertEqual(len(list(index_page)), 2)
+        self.assertEqual(index_page.has_next_page(), False)
+
+    @inlineCallbacks
+    def test_get_dynamic_contact_keys_for_group(self):
+        """
+        If we ask for the dynamic keys for a group, we get a PaginatedSearch
+        that we can walk until we have all the results.
+        """
+        store = self.contact_store
+        group = yield store.new_smart_group(u'test group', u'surname:"Foo 1"')
+        matching_contact = yield store.new_contact(
+            name=u'Contact', surname=u'Foo 1', msisdn=u'12345')
+        yield store.new_contact(
+            name=u'Contact', surname=u'Foo 2', msisdn=u'12345')
+
+        index_page = yield store.get_dynamic_contact_keys_for_group(group)
+        self.assertEqual(list(index_page), [matching_contact.key])
+        self.assertEqual(index_page.has_next_page(), True)
+
+    @inlineCallbacks
+    def test_get_contact_keys_for_group_static(self):
+        """
+        If we ask for the keys for a static group, we get an IndexPage that we
+        can walk until we have all the results.
+        """
+        store = self.contact_store
+        group = yield store.new_group(u'test group')
+        contact_keys = set([])
+        for i in range(2):
+            contact = yield store.new_contact(
+                name=u'Contact', surname=u'%d' % i, msisdn=u'12345',
+                groups=[group])
+            contact_keys.add(contact.key)
+
+        index_page = yield store.get_contact_keys_for_group(group)
+        self.assertEqual(len(list(index_page)), 2)
+        self.assertEqual(index_page.has_next_page(), False)
+
+    @inlineCallbacks
+    def test_get_contact_keys_for_dynamic_group(self):
+        """
+        If we ask for the keys for a dynamic group, we get a wrapper around a
+        PaginatedSearch and an empty IndexPage that we can walk until we have
+        all the results.
+        """
+        store = self.contact_store
+        group = yield store.new_smart_group(u'test group', u'surname:"Foo 1"')
+        matching_contact = yield store.new_contact(
+            name=u'Contact', surname=u'Foo 1', msisdn=u'12345')
+        yield store.new_contact(
+            name=u'Contact', surname=u'Foo 2', msisdn=u'12345')
+
+        first_page = yield store.get_contact_keys_for_group(group)
+        self.assertEqual(list(first_page), [matching_contact.key])
+        # This is the empty last page of the search results.
+        second_page = yield first_page.next_page()
+        self.assertEqual(list(second_page), [])
+        # This is the empty only page of the index results.
+        third_page = yield second_page.next_page()
+        self.assertEqual(list(third_page), [])
+        self.assertEqual(third_page.has_next_page(), False)
```

Comment:
Would it be worth testing the case where we have a non-empty index page (if the smart group had static contacts)?